### PR TITLE
Implements FROM, SELECT, WHERE, NEW

### PR DESCRIPTION
**Pull request by cromwellian**
Adds new Sequence based functions to the Expression DSL: QualifiedExpression and GlobalFunctions

QualifiedExpressions are:
1. FROM - maps over Sequence in the currentScope (source), assigning to iterationVar
2. SELECT/NEW - constructs a new scope using schemaNames and a list of fieldName -> expression mapping over
    the qualifier
3. WHERE - evaluates a boolean, filters over the qualifier

GlobalFunctions - top level functions which consume Sequences and return any kind of value

### CODE REVIEW

**Comment by cromwellian** [comment]
After looking at how I'd implement groupBy, orderBY, which require terminal operations on sequences, I'm going to change the DSL layout and evaluation, i'll have another PR ready sometime this afternoon.

File: javatests/arcs/core/data/expression/ExpressionTest.kt [hunk 1]
```diff
@@ -154,6 +176,113 @@ class ExpressionTest {
         assertThat(evalExpression<Number, Number>(expr, currentScope, "arg" to 1)).isEqualTo(28)
     }
 
+    @Test
+    fun test_evalutor_paxel_from() {
+        val fromExpr = from<Number>("p") on "numbers"
+        assertThat(
+            evalExpression<Sequence<Number>, Sequence<Number>>(fromExpr, currentScope).toList()
```
Comment (piotrswigon) [inline javatests/arcs/core/data/expression/ExpressionTest.kt:183]: The signature of evalExpression is:
```
fun <T, R> evalExpression(
    expression: Expression<T>,
    currentScope: Expression.Scope = mapOf<String, Any>().asScope(),
    vararg params: Pair<String, Any>
): R
```

I am wondering, when would `T` be different than `R` ? If expression is of type `T`, would the result of evaluating it always be of type `T` as well? Or maybe some `T or Error` thing?

File: javatests/arcs/core/data/expression/ExpressionTest.kt [hunk 2]
```diff
@@ -154,6 +176,113 @@ class ExpressionTest {
...
+            (currentScope["p"].asNumber() gt 6.asExpr())
+        val unionExpr = union(lessThan8, greaterThan6)
+
+        assertThat(
+            evalExpression<Sequence<Number>, Sequence<Number>>(unionExpr, currentScope).toList()
+        ).isEqualTo(numbers)
+    }
+
+    @Test
+    fun test_evaluator_paxel_expression() {
```
Comment (piotrswigon) [inline javatests/arcs/core/data/expression/ExpressionTest.kt:256]: This is super exciting!

File: java/arcs/core/data/expression/Evaluator.kt [hunk 3]
```diff
@@ -53,6 +53,110 @@ class ExpressionEvaluator(
...
+        if (sequence is List<*>) {
+            sequence = sequence.asSequence()
+        }
+
+        require(sequence is Sequence<*>) {
+            "${expr.source} of type ${sequence::class} cannot be converted to a Sequence"
+        }
+
+        return sequence.map { value ->
+            currentScope.set(expr.iterationVar, value as Any)
```
Comment (piotrswigon) [inline java/arcs/core/data/expression/Evaluator.kt:71]: Should this be a mutable scope, or a hierarchy of scopes?

The place where mutable scope could bite us is temporary variables and sub-queries:
```
from p in stuff
let max = Max(from x in p.numbers select x)
select new Thing {
  a: max,
  b: max / 2,
  c: max > 29,
  d: x + 5 // <- that definitely not allowed!
}
```
Clearly the scope where `x` is available is very limited. WDYT?

If you think that in the long term we should do it hierarchically, but you want to do it in a separate PR I'm ok with this as well.

File: java/arcs/core/data/expression/Builders.kt [hunk 4]
```diff
@@ -165,15 +171,74 @@ fun <E : Expression.Scope, T> Expression.FieldExpression<E, T>.asScope() =
     this as Expression.FieldExpression<E, Expression.Scope>
 
 /** Constructs a reference to a current scope object for test purposes */
-class CurrentScope<V>(map: Map<String, V>) : MapScope<V>("<this>", map)
+class CurrentScope<V>(map: MutableMap<String, V>) : MapScope<V>("<this>", map)
```
Comment (piotrswigon) [inline java/arcs/core/data/expression/Builders.kt:174]: Will we ever have scopes that are of a single Type? In other words should we skip the `V` type argument and use `Any` explicitly? Would save you some type casting.

File: java/arcs/core/data/expression/Evaluator.kt [hunk 5]
```diff
@@ -53,6 +53,110 @@ class ExpressionEvaluator(
...
+            value
+        }
+    }
+
+    override fun <T> visit(expr: Expression.WhereExpression<T>): Any {
+        return (expr.qualifier.accept(this) as Sequence<T>).filter {
+            expr.expr.accept(this) == true
```
Comment (piotrswigon) [inline java/arcs/core/data/expression/Evaluator.kt:78]: I was baffled by you not using `it` in the filter statement. That's because the expression evaluates in the scope, which is different every time, right?

Another potential reason to avoid mutable scopes - maybe someday we could parallelize operations like `where` and `select`... though I don't think we should worry about this now...

File: java/arcs/core/data/expression/Evaluator.kt [hunk 6]
```diff
@@ -53,6 +53,110 @@ class ExpressionEvaluator(
...
+}
+
+/** Global functions are invoked by [FunctionExpression] during evaluation. */
+sealed class GlobalFunction(val name: String) {
+
+    /** Functions accept a varargs list of [Sequence] and return any type. */
+    abstract fun invoke(args: List<Any>): Any
+
+    /** Performs [Sequence.union] of two [Sequence]s. */
+    object Union : GlobalFunction("union") {
```
Comment (piotrswigon) [inline java/arcs/core/data/expression/Evaluator.kt:116]: Do you have strong preference wrt casing? CamelCase may be better if we ever have functions that have multi-word names. Though you could do under_scopes there. 

I would argue for using the same format here and in the parser, just to limit weirdness. WDYT?

File: java/arcs/core/data/expression/Expression.kt [hunk 7]
```diff
@@ -324,6 +346,80 @@ sealed class Expression<T> {
...
+    /**
+     * Represents an iteration over a [Sequence] in the current scope under the identifier [source],
+     * placing each member of the sequence in a scope under [iterationVar] and evaluating
+     * [iterationExpr], returning a new sequence.
+     *
+     * @param E the type of the [qualifier] expression if any
+     * @param T the type of elements in the resulting [Sequence]
+     */
+    data class FromExpression<E, T>(
+        override val qualifier: Expression<Sequence<E>>?,
```
Comment (piotrswigon) [inline java/arcs/core/data/expression/Expression.kt:364]: Should FromExpression be a QualifiedExpression? It seems that `qualifier` is null, while it is present in WhereExpression and SelectExpression.

Comment (piotrswigon) [reply java/arcs/core/data/expression/Expression.kt:364]: Though I guess we could someday allow:
```
from x in [subquery1, subquery2, subquery3]
(...)
```
And here the `qualifier` would be present, but not the `source`. Though you could likely do something similar with other language constructs...

File: javatests/arcs/core/data/expression/ExpressionTest.kt [hunk 8]
```diff
@@ -154,6 +176,113 @@ class ExpressionTest {
...
+        ).isEqualTo(listOf(5))
+    }
+
+    @Test
+    fun test_evalutor_paxel_max() {
+        val selectMaxExpr = from<Number>("p") on "numbers" select max(currentScope["numbers"])
+
+        assertThat(
+            evalExpression<Sequence<Number>, Sequence<Number>>(selectMaxExpr, currentScope).toList()
+        ).isEqualTo(numbers.map { numbers.size })
```
Comment (piotrswigon) [inline javatests/arcs/core/data/expression/ExpressionTest.kt:209]: you want `numbers.map { numbers.max() }` I think, it just so happens that max == size

File: java/arcs/core/data/expression/Builders.kt [hunk 9]
```diff
@@ -165,15 +171,74 @@ fun <E : Expression.Scope, T> Expression.FieldExpression<E, T>.asScope() =
...
+
+/** Helper to construct [NewExpression]. */
+data class NewBuilder<E, T>(val schemaName: String) {
+    operator fun invoke(
+        block: () -> List<Pair<String, Expression<*>>>
+    ): Expression<T> = Expression.NewExpression(setOf(schemaName), block())
+}
+
+/** Constructs a [NewBuilder] for the given [schemaName]. */
+fun <E, T> new(schemaName: String) = NewBuilder<E, T>(schemaName)
```
[the remark this answers is not in the document]
Comment (cromwellian) [reply java/arcs/core/data/expression/Builders.kt:225]: Done

File: javatests/arcs/core/data/expression/ExpressionTest.kt (hunk 1 above)
Comment (cromwellian) [reply javatests/arcs/core/data/expression/ExpressionTest.kt:183]: Done

File: java/arcs/core/data/expression/Evaluator.kt (hunk 3 above)
Comment (cromwellian) [reply java/arcs/core/data/expression/Evaluator.kt:71]: It should be a hierarchy (scope chain). Let me do this in a followup since it could be a big change.

File: java/arcs/core/data/expression/Evaluator.kt (hunk 5 above)
Comment (cromwellian) [reply java/arcs/core/data/expression/Evaluator.kt:78]: it's essentially doing. collection.filter { scope.lookup("p") < 5 }, because the Sequence created by the qualifier (in this case, the FROM expression) is setting "p", it doesn't refer to it. This is a consequence of linq, e.g. you could do

from p in collection
where random() < foo

so 'it' isn't involved, but may it, in which case, it appears as 'p' in the scope already

File: java/arcs/core/data/expression/Expression.kt (hunk 7 above)
Comment (cromwellian) [reply java/arcs/core/data/expression/Expression.kt:364]: You can do this
```
from x in p
from y in p.q
where y.blah == "Hello"
```

In which case, this is 
```
WHERE(FROM(FROM("x", "p"), "y", currentScope["p"]["q"]), currentScope["y"]["blah"] eq "Hello")
```

File: javatests/arcs/core/data/expression/ExpressionTest.kt (hunk 8 above)
Comment (cromwellian) [reply javatests/arcs/core/data/expression/ExpressionTest.kt:209]: Done

File: java/arcs/core/data/expression/Builders.kt (hunk 4 above)
Comment (cromwellian) [reply java/arcs/core/data/expression/Builders.kt:174]: I tried removing this but it led to some other problems. I'll remove in follow ons

File: java/arcs/core/data/expression/Evaluator.kt (hunk 6 above)
Comment (cromwellian) [reply java/arcs/core/data/expression/Evaluator.kt:116]: You mean you want the function to be case sensitive "Union"? That's more a C# rule though, because in Java/Kotlin, it would begin with lowercase char?

I don't have a preference, whatever is preferred.
